### PR TITLE
ci: pin tagged release publication inputs

### LINKED ISSUE #2387: Tagged release re-reads live ci-metadata and per-leg ports HEAD instead of the build pin
## Summary

Nightly pins `matrix_sha` and fail-closes if `ci-metadata` moves. Tagged publish does **not**.

Build bakes `matrix_row` from **today’s** `release_matrix`. Catalog publish re-reads **live** `origin/ci-metadata` via `--print-route` and requires **byte-exact** row equality (axis 7), plus one asset per current build-role row.

`--print-route` is raw `.versions` (`read-version-matrix.sh`). Release rows are rewritten (`extra_pkgs` defaulted/sorted, `.arch`/`.ci`/`.mac` dropped). `_normalize_route_matrix` only pops `ci`.

Notes can sit in draft for days. Any 2.8 add/remove/status/`extra_pkgs`/`image_name` tweak after the cut fails catalog publish **after** the GitHub Release is already public.

Each matrix leg also `git ls-remote`s `pfblockerng/use-github` at build time (`release.yml` ~1145–1217). Nightly resolves one ports SHA in `prepare`. Mid-fan-out push (including an earlier draft’s `sync-ports-fork`) can bake **different** ports SHAs / recipes into CE 2.8 vs Plus of the same tag.

## Fix

- Persist the build-time ROUTE JSON (artifact / draft output). Publish and republish must use that pin, same as Nightly.
- Resolve `PORTS_SHA` once in `prepare-release` / `read-matrix`; pass that SHA to every leg.

## Smoke / CI that would have caught this

**Contract tests (no VM):**

1. `tests/test_publish_catalogues.py`: bake records from matrix M1; republish with live `--print-route` = M1 plus a mutated `extra_pkgs` / dropped 2.8 row → must use **baked** rows, not fail closed on live drift (or must refuse publish *before* the GitHub Release exists — pick one and pin it).
2. `tests/test_release_ci_path_consistency.py`: every `build-leg` invocation in `release.yml` receives `PORTS_SHA` from prepare, not an inline `git ls-remote`.

**Release dry-run:** two-leg matrix; mutate `ci-metadata` between `build-pkgs-portable` and `publish-pkg-repo`; catalog publish must still write `ce-2.8` from the baked record (or must have been blocked at draft, not after `release: published`).

### PR DESCRIPTION
**Pull request by andrebrait**
## Summary

- resolve one `ci-metadata` SHA, ROUTE matrix, and FreeBSD Ports SHA before tagged build fan-out
- attach `pfblockerng-release-handoff.json` to the draft and consume it from both publication callbacks
- authenticate the handoff against the GitHub Release API asset digest, then validate exact release tag, source SHA, Ports SHA, and every canonical build record before catalogue mutation
- leave Nightly and the publisher state machine unchanged

## Evidence

- Initial production change: frozen focused tests failed on untouched `devel` for the missing handoff and live-input paths, then passed unchanged (`80 passed`; POSIX shellspec `73 examples, 0 failures`).
- Initial RED-time hashes, unchanged through production commit `ef299edcab569cf5c33f3c033f1745a9d47ebc0f`:
  - `tests/test_tagged_release_handoff.py` `1c8489015e9eab2e3045815ff1fc94fc19b2fbab`
  - `tests/test_release_ci_path_consistency.py` `0a7f98274c61eb87a528c573108f8b42a8c7e866`
  - `tests/test_issue2143_republish_workflow.py` `6c99aa6decbc1209301ff6a3e0d36b008e6b79ea`
  - `tests/test_publish_release.py` `9b07e3a3cadf899014d07a2f8286f5d50d1297d1`
  - `tests/shell/publish_pkg_repo_spec.sh` `319eab947e285cb9d5ad01b7016bb28414664fcf`
- Review fix: strict integer schema cases failed before the production edit and passed after (`2 passed`) with frozen hash `ad1bbbc8a198ff41a861919ea7cee0f6b480bbb0`.
- Review hardening: callback fixture isolation kills a missing-download mutation; safe workflow `env:` wiring failed before the production edit and passed after (`2 passed`) with frozen hash `a6eb910df2508f8190cf2cfc3daeb6d3af9888c7`.
- Final head `9dbffcd9fc2bbcc5490f21cbd3a1fe91bad0d3a6` regression hashes:
  - `tests/test_tagged_release_handoff.py` `a70fbc5cf69aa8682a8ab527e33272e8f61cc54c`
  - `tests/test_release_ci_path_consistency.py` `a6eb910df2508f8190cf2cfc3daeb6d3af9888c7`
  - `tests/test_issue2143_republish_workflow.py` `153ebfc3c470ca98064ee9b2a7dc541c529c56a0`
  - `tests/test_publish_release.py` `4de43d50416dcf30343030b20ea0a96ec12ee598`
  - `tests/shell/publish_pkg_repo_spec.sh` `319eab947e285cb9d5ad01b7016bb28414664fcf`
- Final exact-head canonical gates: `GATES: PASS` (`uv run --locked pytest`, ruff lint/format, mypy, `sh -n`, ShellCheck, dash shellspec).

Fixes #2387

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

### CODE REVIEW

**Comment by andrebrait** [comment]
## Adversarial review — contract conformance

- Head reviewed: `023ab1b203b18985a399d4ad9f45342012ccc9e9`
- Model: `gpt-5.6-terra` (mid-tier)
- Effort: OMP reviewer surface did not expose an introspectable effort value; required contract-review setting is `medium`
- Verdict: **CLEAN** — blocking 0, nitpick 0

Coverage passed: one pre-fan-out `ci-metadata`/ROUTE/Ports pin; one durable draft handoff; automatic and manual callbacks require and authenticate that handoff; exact tag/source/Ports/build-record binding before mutation; malformed/missing/inconsistent input fails closed; Nightly and sibling state-machine work remain untouched. All 14 changed files were examined. Focused contract run: `20 passed`; broader changed-file run: `177 passed`.

Finding ledger: no prior or new contract findings.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Adversarial review — correctness and hostile inputs

- Head reviewed: `023ab1b203b18985a399d4ad9f45342012ccc9e9`
- Model: `gpt-5.6-sol` (top-tier)
- Effort: `high`
- Verdict: **CLEAN**

Focused head proof: `19 passed`. Per-file review covered the handoff schema/loader, both callback workflows, release pinning, publisher validation, shell forwarding, and all changed tests. Missing, duplicate, malformed, wrong-tag, wrong-source, wrong-Ports, inconsistent-record, empty-route, schema-boolean, and schema-float inputs were exercised.

Finding ledger:
1. `scripts/tagged_release_handoff.py`: JSON `true`/`1.0` could satisfy `schema == 1` — **fixed** at `af16ec7cebe64bb689a045fc16b45c0320579266`; frozen RED cases now require an exact integer.
2. New findings at final head — none.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Adversarial review — over-engineering

- Head reviewed: `023ab1b203b18985a399d4ad9f45342012ccc9e9`
- Model: `gpt-5.6-sol` (top-tier)
- Effort: `high`
- Verdict: **CLEAN for merge** — blocking 0; optional cuts only

Finding ledger:
1. Self-authenticated `route_matrix_sha256` duplicated the GitHub asset digest — **fixed**; field removed.
2. Handcrafted publisher fixture duplicated the production constructor — **fixed**.
3. Two workflows duplicate the handoff-download body — **skipped**: two explicit workflow boundaries are clearer and a shared shell helper would add coupling for no net behavior.
4. Shell malformed-handoff case overlaps loader/failure containment — **skipped**: it pins the observable wrapper boundary, not the loader implementation.
5. Test module loader repeats a few lines — **skipped**: isolated loads preserve test independence; module-global caching would save about five lines but change fixture semantics.

Estimated optional removable lines: ~49; accepted removals: self-digest and duplicate fixture machinery. No speculative production abstraction remains.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Adversarial review — test honesty

- Head reviewed: `023ab1b203b18985a399d4ad9f45342012ccc9e9`
- Model: `gpt-5.6-luna` (small-tier)
- Effort: OMP reviewer surface did not expose an introspectable effort value; required Luna review setting is `high`
- Verdict: **CLEAN**

Changed-file run: `177 passed`. Targeted mutations proved the tests reject: a vacuous handoff-count guard; wrong `ci-metadata` fetch/rev-parse refs; removed callback handoff download; wrong handoff Ports identity; reintroduced self-authenticated route digest; and wrong callback release/repository arguments.

Finding ledger:
1. Missing/duplicate handoff healthcheck fixture was vacuous — **fixed**.
2. Route digest correctness assertion was circular — **fixed** by removing the self-authenticated field and pinning the exact output field set.
3. Pin-step fake git did not bind the fetched ref — **fixed**.
4. Callback checks were source-text-only and ignored download identity — **fixed** with executable fake-`gh` checks and exact call assertions.
5. Publisher tests handcrafted all handoffs — **fixed** by using the production constructor.
6. Shell argument-boundary gap — **fixed** at final head.

All changed tests carry regression-killing assertions; no weakened assertions or vacuous negatives remain.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
@coderabbitai review

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

File: tests/test_issue2143_republish_workflow.py [hunk 1]
```diff
@@ -129,3 +153,98 @@ def test_manual_republish_rejects_release_selector_before_api(tmp_path: Path) ->
...
+                    "HANDOFF_SOURCE": str(handoff),
+                    "META_FILE": str(meta),
+                    "GH_LOG": str(gh_log),
+                    "RUNNER_TEMP": str(runner_temp),
+                },
+                capture_output=True,
+                text=True,
+                check=False,
+            )
+            assert (completed.returncode == 0) == (expected == 0), completed.stdout + completed.stderr
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/test_issue2143_republish_workflow.py]: APPLY — fixed in `9dbffcd9fc2bbcc5490f21cbd3a1fe91bad0d3a6`.

Confirmed before the edit by removing only the manual callback's handoff download: the test incorrectly passed (`1 passed`). `RUNNER_TEMP` now includes `job_name`; the same mutation fails on the missing handoff digest, while the unmodified test passes.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

File: tests/test_tagged_release_handoff.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            "--route-matrix",
+            str(route),
+            "--output",
+            str(output),
+        ],
+        cwd=ROOT,
+        capture_output=True,
+        text=True,
+        check=False,
+    )
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/test_tagged_release_handoff.py:81]: SKIP — wrong execution-surface premise.

This file is a host-side pytest, not appliance code. `.agents/context/lang-python.md` forbids direct Python invocation **on the appliance** and explicitly permits bare developer/CI Python under `scripts/`; the package's appliance consumers remain routed through `pfb_python.sh`. Calling the module directly would stop testing the shipped CLI boundary and its argument parsing.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
## Review orchestrator — final ledger

Head: `9dbffcd9fc2bbcc5490f21cbd3a1fe91bad0d3a6`

### Findings

1. Handoff schema accepted JSON booleans/floats as integer schema 1 — **fixed@af16ec7cebe64bb689a045fc16b45c0320579266**; frozen RED→GREEN hash `ad1bbbc8a198ff41a861919ea7cee0f6b480bbb0`.
2. Callback healthcheck fixture always injected one handoff — **fixed@af16ec7cebe64bb689a045fc16b45c0320579266**.
3. Self-authenticated route digest and circular assertion — **fixed@af16ec7cebe64bb689a045fc16b45c0320579266**; GitHub Release API authenticates the whole asset.
4. Pin-step fake git did not bind the fetched ref — **fixed@af16ec7cebe64bb689a045fc16b45c0320579266**.
5. Callback checks ignored exact release/repository download identity — **fixed@023ab1b203b18985a399d4ad9f45342012ccc9e9**.
6. CodeRabbit: callback workflows reused `RUNNER_TEMP` — **fixed@9dbffcd9fc2bbcc5490f21cbd3a1fe91bad0d3a6**; removing only the second workflow's handoff download passed before and fails after.
7. CodeRabbit: direct `${{ ... }}` interpolation in the write-capable handoff step — **fixed@9dbffcd9fc2bbcc5490f21cbd3a1fe91bad0d3a6**; identities now enter through quoted environment variables. Frozen RED→GREEN hash `a6eb910df2508f8190cf2cfc3daeb6d3af9888c7`.
8. CodeRabbit: host pytest uses `sys.executable` — **skipped**: the cited rule forbids direct Python only on the appliance; this host/CI test intentionally exercises the shipped CLI boundary. Appliance consumers remain behind `pfb_python.sh`.
9. Optional ponytail cuts (two-workflow helper, shell boundary case, module loader) — **skipped**: each suggested abstraction/deletion either adds coupling or removes a distinct observable boundary; no blocking speculative production code remains.

### Gates

- Four adversarial legs complete; final outcomes recorded in their audit comments.
- CodeRabbit finished on `023ab1b2`; every inline and summary finding above is resolved. Its concrete mechanical fixes do not trigger a second spend.
- Final exact-head canonical gates: `GATES: PASS`.
- Final exact-head CI: `PASS`; all required checks green. Benchmark/UI fan-out jobs are intentionally skipped by path scope.
- Catch-all sweep found no untriaged review, inline, or top-level comment.

🤖 Generated by Oh My Pi and posted on behalf of @andrebrait.
